File: Data Structures & Algorithms/valid-tree/submission-0.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        if (n - 1 != edges.size()) return false;

        vector<vector<int>> graph(n);

        for (auto& e : edges) {
            graph[e[0]].push_back(e[1]);
            graph[e[1]].push_back(e[0]);
        }

        vector<bool> visited(n, false);

        function <bool(int, int)> dfs = [&] (int node, int parent) -> bool {
            visited[node] = true;

            for (int nei : graph[node]) {
                if (nei == parent) continue;
                if (visited[nei]) return true;
                if (dfs(nei, node)) return true;
            }

            return false;
        };

        if (dfs(0, -1)) return false;

        for (bool v : visited) {
            if (!v) return false;
        }

        return true;
    }
};

File: Data Structures & Algorithms/redundant-connection/submission-0.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        mp = defaultdict(list)
        nodes = set()
        for u, v in edges:
            mp[u].append(v)
            mp[v].append(u)
            nodes.add(u)
            nodes.add(v)

        def traverse(mp, x, y):
            visited = set()

            level = deque([x])

            while level:
                n = len(level)

                for i in range(n):
                    node = level.popleft()
                    if node in visited:
                        continue
                    visited.add(node)
                    
                    neighbors = mp[node]

                    for neigh in neighbors:
                        if (node == x and neigh == y) or (node == y and neigh == x):
                            continue
                        level.append(neigh)
                        
            return visited

        res = []
        for u, v in edges:
            print(traverse(mp, u, v))
            if len(nodes) == len(traverse(mp, u, v)):
                res.append([u, v])
        
        for i in range(len(edges) - 1, -1, -1):
            if edges[i] in res:
                return edges[i]